baseplate/context/cassandra.py

```diff
@@ -90,7 +96,50 @@ def make_object_for_context(self, name, server_span):
         return cqlmapper.connection.Connection(session_adapter)
 
 
-def _on_execute_complete(_, span):
+class WaitForCallbackResponseFuture(object):
+    """Wrap the ResponseFuture to ensure callbacks have completed.
+
+    This fixes a race condition where the server span can complete before
+    the callback has closed out the child span.
+
+    """
+
+    def __init__(self, span, future, callback_fn, callback_args, errback_fn, errback_args):
+        self.callback_event = Event()
+
+        def register_completion(fn):
+            def register_after_execute(*a, **kw):
+                try:
+                    return fn(*a, **kw)
+                finally:
+                    self.callback_event.set()
+            return register_after_execute
+
+        future.add_callback(register_completion(callback_fn), callback_args)
+        future.add_errback(register_completion(errback_fn), errback_args)
+
+        self.future = future
+
+    def result(self):
+        exc = None
+
+        try:
+            result = self.future.result()
+        except Exception as e:
+            exc = e
```

Comment:
This will lose the original traceback. According to [this](http://www.ianbicking.org/blog/2007/09/re-raising-exceptions.html) we can do:

```python
try:
    ...
except Exception:
    exc_info = sys.exc_info()

...

if exc_info:
    raise exc_info[0], exc_info[1], exc_info[2]
```
which will preserve that even though we're reraising non-locally

Reply:
hmm, this doesn't seem to work in python3. I'll hunt around for a solution